Post-process results to standard format and provide generic results printing command

johnSchnake opened the pull request:
**What this PR does / why we need it**:
Adds a post-processing step where we walk the files the plugin produced and dump info into a single file whose format we define.

For junit, we record each tests and it's pass/fail status. For raw files (since we may not know how large they are) we just record each file location.

This file (which now contains all tests/output files) can be parsed by clients in order to more easily view plugin results.

The plugin manifests have new fields `ResultFormat` and `ResultFile` to inform this post-processing step.

**Which issue(s) this PR fixes**
Fixes #306

**Special notes for your reviewer**:

**Release note**:
```
TBD
```

johnSchnake commented on Makefile:
```diff
@@ -24,7 +24,7 @@ REGISTRY ?= gcr.io/heptio-images
 DIR := ${CURDIR}
 DOCKER ?= docker
-LINUX_ARCH := amd64 arm64
+LINUX_ARCH := amd64
```
Dont commit this; it is just to iterate faster so it doesnt build the others.

johnSchnake commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,167 @@
+	return nil
+}
+
+func printResultsDetails(o *results.Item, input resultsInput) {
```
All these `result` output stuff is tbd. It wasn't too terribly defined in the design doc (focus was on server side stuff). We can iterate a bit on it after we finalize the other stuff.

johnSchnake commented on cmd/sonobuoy/app/root.go:
```diff
@@ -53,6 +53,9 @@ func NewSonobuoyCommand() *cobra.Command {
 	cmds.AddCommand(NewCmdImages())
+	cmds.AddCommand(NewCmdResults())
+	// Debug only
+	cmds.AddCommand(NewSchnake())
```
Don't commit; just my own debug command.

johnSchnake commented on pkg/client/results/doc.go:
```diff
@@ -17,3 +17,22 @@ limitations under the License.
+// Item is the central format for plugin results. Various plugin
+// types can be transformed into this simple format and set at a standard
+// location in our results tarball for simplified processing by any consumer.
+type Item struct {
```
Not sure why this ended up in `doc.go`; I'll move it to somewhere more appropriate.

johnSchnake commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,169 @@
+)
+
+type postProcessor func(string) (Item, error)
+type fileSelector func(string, os.FileInfo) bool
```
Just interesting to note that github doesn't properly do syntax highlighting here.

johnSchnake commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,169 @@
+func aggregateStatus(items ...Item) string {
+	for _, v := range items {
+		if len(v.Items) > 0 {
+			v.Status = aggregateStatus(v.Items...)
```
This will definitely need tests and currently doesn't either work as intended or wasn't called in the right place. Last I checked I had some empty values that shouldn't have been empty.

johnSchnake commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,169 @@
+	fmt.Printf("checking types %+v", p, p.GetResultFormat())
+	switch p.GetResultFormat() {
+	case "junit", "e2e":
+		fmt.Println("got junit item")
```
Lots of debug items thoughout; reminder to remove them all.

johnSchnake commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,169 @@
+
+func processPluginWithProcessor(p plugin.Interface, baseDir string, processor postProcessor, selector fileSelector) (Item, error) {
+	pdir := path.Join(baseDir, PluginsDir, p.GetName())
+	_, isDS := p.(*daemonset.Plugin)
```
Just special handling down below for daemonsets; default to the other handling for jobs and any "other" types that could in theory exist (since the someone using this as a library could be making their own type for the plugin interface).

johnSchnake commented on pkg/client/results/processing_test.go:
```diff
@@ -0,0 +1,64 @@
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			item, err := ProcessRawPlugin(tc.plugin, tc.dir)
```
Originally written when this method existed; I dont think it even does any more.

johnSchnake commented on pkg/client/results/raw.go:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	return Item{
+		Name:   rawPath,
```
Make this `filepath.Base` of the path. Even though it would be duplicate info, may want to add the `metadata.file` with the full path to be consistent with how the junit results set that field.

johnSchnake commented on pkg/client/results/reader.go:
```diff
@@ -295,3 +297,57 @@ func ConfigFile(version string) string {
+				q.Q("found", wantPath) // DEBUG
+			}
+			q.Q(curPath, wantPath) // DEBUG
+			return ExtractBytes(wantPath, curPath, info, &buf)
```
Delete old code

johnSchnake commented on pkg/discovery/discovery.go:
```diff
@@ -241,6 +258,117 @@ func getPodLogNamespaceFilter(cfg *config.Config) string {
+			resultObj.Status = "failed"
+			status = "failed"
+		case t.Skipped != nil:
+			status = "skipped"
```
Should be using constants

johnSchnake commented on pkg/plugin/aggregation/run_test.go:
```diff
@@ -239,6 +239,10 @@ func (cp *MockCleanupPlugin) SkipCleanup() bool {
 
+func (cp *MockCleanupPlugin) GetResultFile() string {
+	return ""
+}
```
Just necessary to add since I added the new method to the interface.

Trivial for people who want to update the library to add a similar method.

johnSchnake commented on pkg/client/results/junit.go:
```diff
@@ -52,3 +57,46 @@ func Passed(testCase reporters.JUnitTestCase) bool {
+	resultObj := Item{
+		Name:     filepath.Base(junitPath),
+		Status:   StatusPassed,
+		Metadata: map[string]string{"file": junitPath},
```
Be sure to add metadata to the raw output stuff too

johnSchnake replied:
Should be relative path instead of full path?

johnSchnake commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,167 @@
+			// load file and print it
+			r, cleanup, err := getReader(input.archive)
+			if err != nil {
+				fmt.Println("Schnake err", err)
```
logrus

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,188 @@
+	defer cleanup()
+	if err != nil {
+		errlog.LogError(err)
+		os.Exit(1)
```
This error should be returned so that the exit can be handled by the `Run` function in the command.

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,188 @@
+		os.Exit(1)
+	}
+
+	obj, err := r.PluginResults(input.plugin)
```
This error isn't being checked/returned.

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,188 @@
+	return nil
+}
+
+func printResultsDetails(prefix string, o *results.Item, input resultsInput) {
```
If you wanted to test this functionality a bit more easily, maybe you could pass a writer to this function which the formatted output could be written to rather than trying capture what is on stdout.

johnSchnake replied:
Yea, I could do that. I opted to just use `Example...` tests which tests stdout output automatically. I lost the ability to do a table driven test, but for just testing formatting of a CLI command I think that is OK.

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,188 @@
+			}
+			if path == filepath {
+				found = true
+				freader := info.Sys().(io.Reader)
```
Should we be checking that the type assertion was successful here?

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	switch o.Status {
+	case "passed":
```
The `Status*` constants could be used here.

zubron commented on pkg/client/results/junit.go:
```diff
@@ -52,3 +58,51 @@ func Passed(testCase reporters.JUnitTestCase) bool {
+
+	infile, err := os.Open(currentFile)
+	if err != nil {
+		resultObj.Status = StatusFailed
```
I'm not sure if the status should be `Failed` if we fail to open the file or decode it as there could be non-test related reasons e.g. incorrect file permissions or something. Perhaps `StatusUnknown` would work for these cases? I don't feel very strongly about it though, so if you had particular reasons for using that status that's fine.

zubron commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,181 @@
+	var err error
+
+	switch p.GetResultFormat() {
+	case "junit", "e2e":
```
Perhaps these could be const values? Also, will there be any list of supported result types?

johnSchnake replied:
Added constants; didnt exactly add a "list" explicitly, but the supported types are just the ones that we have constants for. Our documentation will reference those.

johnSchnake commented on pkg/client/results/reader.go:
```diff
@@ -295,3 +297,57 @@ func ConfigFile(version string) string {
+				return err
+			}
+			// TODO(chuckha) consider reusing this function for any generic e2e-esque plugin results.
+			// TODO(chuckha) consider using path.Join()
```
The copied code was just used as the starting place; comments removed.

The code is now used in a general sense to get any result file for a given plugin.

johnSchnake replied:
I just left the original code/comment alone since it is tied into the `e2e` command which has a bunch of other things that make it unique.

johnSchnake commented:
Codecov doesnt seem to handle force pushing onto branches too well; hard to navigate to see new coverage.

Coverage details here: https://codecov.io/gh/heptio/sonobuoy/commit/2d768fab5a88d6a8cd3ca7b128a65bd215dcec21
Currently at 61% coverage

johnSchnake commented on pkg/plugin/loader/loader.go:
```diff
@@ -66,11 +66,13 @@ func LoadAllPlugins(namespace, sonobuoyImage, imagePullPolicy, imagePullSecrets
 		if err != nil {
 			return []plugin.Interface{}, errors.Wrapf(err, "couldn't load plugin definition file %v", file)
 		}
+		fmt.Printf("Loaded file: %+v\n\n", string(definitionFile))
```
These shouldn't have gotten added; remove

johnSchnake commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,255 @@
+	switch input.mode {
+	case resultModeDetailed:
+		printResultsDetails([]string{}, obj, input)
+	case resultModeDump:
```
Didnt add the logic here; intended to just print the file to stdout

johnSchnake commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+func printResultsSummary(o *results.Item) {
+	fmt.Printf("Plugin: %v\nStatus:%v\n", o.Name, o.Status)
```
missing space after status

johnSchnake commented:
Changes to point out in recent pushes:
 - modified json serialization of `Item` type to omitempty on a few fields and use lower case. Tests modified accordingly
 - modified the `WalkFiles` logic to mirror that of typical `filepath.WalkFunc` logic: processing should stop when an error is encountered. This allows us to stop processing when we find the file we want and we can return an `io.Reader`. This allows us to (1) reuse that logic in a few places (2) be more efficient instead of loading the whole item into a byte slice.
 - modified some names in a few test cases to be more clear
 - added logic for `--mode dump` to dump the whole results file
 - fixed issue where the json objects in a junit, detailed result set had their own name as part of their `metadata.path` value (duplicating the value in the `name` field)

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,240 @@
+		o.Metadata["path"] = strings.Join(treePath, "|")
+		b, err := json.Marshal(o)
+		if err != nil {
+			logrus.Errorf("Error marshalling item to json: %v", err)
```
By logging these errors rather than returning them and letting them bubble up, the error from `result` will be `nil` which will affect the overall exit code. Is this intended?

johnSchnake replied:
meh, I dont think I spent that much time thinking about it. I'll go ahead and just bubble everything up. This only means that if one node is poorly formed then we abort instead of continuing but I dont think that is that big of a deal.

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,240 @@
+		// capabilities.
+		r, cleanup, err := getReader(input.archive)
+		if err != nil {
+			logrus.Errorf("Failed to read archive to get file %v: %v", o.File, err)
```
`getReader` can still return a `cleanup` function when it returns an error. If we return here, we miss the cleanup call below so it should also be made here.

johnSchnake replied:
👍 Also modified the code to always return a non-nil function (and documented it as such) so I don't have to worry about nil checks.

zubron commented on pkg/client/results/testdata/mockResults/plugins/ds-junit-01/ds-junit-01.golden.json:
```diff
@@ -0,0 +1 @@
+{"name":"ds-junit-01","status":"passed","meta":null,"file":"","items":[{"name":"global","status":"passed","meta":null,"file":"","items":[{"name":"output.xml","status":"passed","meta":{"file":"results/global/output.xml"},"file":"","items":[{"name":"[k8s.io] Pods should be submitted and removed [NodeConformance] [Conformance]","status":"passed","meta":null,"file":"","items":null},{"name":"[sig-node ...
```
It looks like `null` fields are still being included in the output. I thought they would have been skipped with the `omitempty` tag.

johnSchnake replied:
I can fix those up, though it doesnt matter. In this situation, the item is same once loaded/compared.

zubron replied:
Ah, sorry. I had a couple of comments which were outdated, one of which applied to the example test which _does_ show the empty fields being omitted. I thought these test files had been regenerated following those `omitempty` changes. Never mind this comment then!

zubron commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,240 @@
+			return
+		}
+		resultFile := path.Join(results.PluginsDir, input.plugin, o.File)
+		filereader, err := r.FileReader(resultFile)
```
This error isn't being checked.

zubron commented on pkg/client/results/processing.go:
```diff
@@ -0,0 +1,210 @@
+	Name     string            `json:"name" yaml:"name"`
+	Status   string            `json:"status" yaml:"status"`
+	Metadata map[string]string `json:"meta,omitempty" yaml:"meta,omitempty"`
+	File     string            `json:"file,omitempty" yaml:"file,omitempty"`
```
Is `File` needed if we're also including the path within `Metadata`? It's only being set when processing raw files. I'm not sure if it should only be included there or if it's being missed from the junit processor.

johnSchnake replied:
So the `File` field is meant to be a breadcrumb for raw files, currently, an item should set either `File` or `Items` but not both.

The metadata field was added later and the idea of putting the full filepath made sense for tests since it wasn't listed elsewhere. You're right that it is now redundant for the `raw` type.

I think you're right that we can remove it, let me work on that change. Instead of checking if `.File == ""` we will just check `if .Metadata!=nil && .Metadata["file"] != ""`. It is a bit more verbose but it makes a single source of truth for the filename.

johnSchnake commented on pkg/client/results/testdata/mockResults/plugins/ds-junit-01/results/global/output2.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Yes I can do that. I generated them all via just a bunch of copy/paste but now that the test cases are done I can make the file contents/names more clear.

johnSchnake commented on pkg/client/results/testdata/mockResults/plugins/ds-junit-01/results/global/output.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+  <testsuite tests="204" failures="0" time="0.0574002">
```
No, I could actually remove the files that dont matter (actual logs/results). The only thing that matters is the summary.

johnSchnake replied:
Sorry, I misunderstood the question.

No it doesn't matter but one day it actually might. For now I'll leave it as is so that it is even discernable where the data comes from.

I don't use those values (I manually iterate over the data) so that I can get passed/failed/skipped whereas this only shows tests (total) and failed.

johnSchnake commented on cmd/sonobuoy/app/results.go:
```diff
@@ -0,0 +1,255 @@
+	data := resultsInput{}
+	cmd := &cobra.Command{
+		Use:   "results archive.tar.gz",
+		Short: "Inspect e2e test results. Optionally rerun failed tests",
```
Great catch, thanks.
